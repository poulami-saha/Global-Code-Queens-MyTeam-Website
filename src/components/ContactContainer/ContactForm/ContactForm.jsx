import React, { useState } from "react";
import styles from "./ContactForm.module.css";
import FormElement from "../../utility/FormElement/FormElement";
import Button from "../../utility/Button/Button";

const ContactForm = () => {
  const initialState = {
    name: {
      value: "",
      error: "",
      touched: false,
      placeholder: "Name",
      type: "text",
    },
    email: {
      value: "",
      error: "",
      touched: false,
      placeholder: "Email Address",
      type: "text",
    },
    companyName: {
      value: "",
      error: "",
      touched: false,
      placeholder: "Company Name",
      type: "text",
    },
    title: {
      value: "",
      error: "",
      touched: false,
      placeholder: "Title",
      type: "text",
    },
    message: {
      value: "",
      error: "",
      touched: false,
      placeholder: "Message",
      type: "textarea",
    },
  };

  const [formData, setFormData] = useState(initialState);

  const handleChange = (inputField) => (e) => {
    const { value } = e.target;
    setFormData((prevState) => ({
      ...prevState,
      [inputField]: {
        ...prevState[inputField],
        value: value,
        error: "",
        touched: true,
      },
    }));
  };

  // check if email is valid
  const isEmailValid = (email) => {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
  };

  const handleBlur = (inputField) => () => {
    if (!formData[inputField].value) {
      setFormData((prevState) => ({
        ...prevState,
        [inputField]: {
          ...prevState[inputField],
          error: "This field is required",
        },
      }));
    } else if (
      inputField === "message" &&
      formData[inputField].value.length < 25
    ) {
      setFormData((prevState) => ({
        ...prevState,
        [inputField]: {
          ...prevState[inputField],
          error: "Message must be at least 25 characters",
        },
      }));
    } else if (
      inputField === "email" &&
      !isEmailValid(formData[inputField].value)
    ) {
      setFormData((prevState) => ({
        ...prevState,
        [inputField]: {
          ...prevState[inputField],
          error: "Invalid email address",
        },
      }));
    }
  };
  const handleFocus = (inputField) => () => {
    setFormData((prevState) => ({
      ...prevState,
      [inputField]: { ...prevState[inputField], error: "", touched: true },
    }));
  };
  const isFormValid = () => {
    return !Object.values(formData).some(
      (field) => field.error || !field.touched
    );
  };
  const handleSubmit = (e) => {
    e.preventDefault();
    console.log("formData" + formData);
    setFormData(initialState);
  };
  return (
    <form className={styles.flexColumnStart} onSubmit={handleSubmit}>
      {Object.entries(formData).map(
        ([key, { value, error, placeholder, type }]) => (
          <FormElement
            key={key}
            type={type}
            placeholder={placeholder}
            value={value}
            onChange={handleChange(key)}
            onBlur={handleBlur(key)}
            onFocus={handleFocus(key)}
            error={error}
          />
        )
      )}
      <Button elType="secondary" disabled={!isFormValid()}>
        submit
      </Button>
    </form>
  );
};

export default ContactForm;