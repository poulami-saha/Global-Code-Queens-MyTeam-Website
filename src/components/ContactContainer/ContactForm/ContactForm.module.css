.flexColumnStart {
  display: flex;
  flex-direction: column;
  align-items: start;
  gap: 24px;
  width: 100%;
}

@media (min-width: 960px) {
  .flexColumnStart {
    width: 100%;
  }
}
