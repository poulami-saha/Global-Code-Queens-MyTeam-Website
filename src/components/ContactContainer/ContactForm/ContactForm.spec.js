import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import ContactForm from "./ContactForm";

describe("FormComponent", () => {
  const originalConsoleLog = console.log;
  const consoleLogMock = jest.fn();

  beforeAll(() => {
    console.log = consoleLogMock;
  });

  afterAll(() => {
    console.log = originalConsoleLog;
  });

  afterEach(() => {
    consoleLogMock.mockClear();
  });

  test("renders ContactForm correctly and matches snapshot", () => {
    const { asFragment } = render(<ContactForm />);

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();

    // Check if form fields are rendered
    const name = screen.getByPlaceholderText("Name");
    const email = screen.getByPlaceholderText("Email Address");
    const company = screen.getByPlaceholderText("Company Name");
    const title = screen.getByPlaceholderText("Title");
    const message = screen.getByPlaceholderText("Message");

    const selectors = [name, email, company, title, message];
    selectors.map((selector) => expect(selector).toBeInTheDocument());
  });

  test("submits Contact Form with Data", () => {
    render(<ContactForm />);
    // Check if form fields are rendered
    const name = screen.getByPlaceholderText("Name");
    const email = screen.getByPlaceholderText("Email Address");
    const company = screen.getByPlaceholderText("Company Name");
    const title = screen.getByPlaceholderText("Title");
    const message = screen.getByPlaceholderText("Message");

    const submitButton = screen.getByText("submit");

    expect(submitButton).toBeDisabled();

    fireEvent.change(name, { target: { value: "Person" } });
    fireEvent.change(email, { target: { value: "test@gmail.com" } });
    fireEvent.change(company, { target: { value: "Test Gmbh" } });
    fireEvent.change(title, { target: { value: "Software Engineer" } });
    fireEvent.change(message, { target: { value: "We need support" } });

    fireEvent.click(submitButton);

    const expected = {
      companyName: {
        error: "",
        placeholder: "Company Name",
        touched: true,
        type: "text",
        value: "Test Gmbh",
      },
      email: {
        error: "",
        placeholder: "Email Address",
        touched: true,
        type: "text",
        value: "test@gmail.com",
      },
      message: {
        error: "",
        placeholder: "Message",
        touched: true,
        type: "textarea",
        value: "We need support",
      },
      name: {
        error: "",
        placeholder: "Name",
        touched: true,
        type: "text",
        value: "Person",
      },
      title: {
        error: "",
        placeholder: "Title",
        touched: true,
        type: "text",
        value: "Software Engineer",
      },
    };

    expect(consoleLogMock).toHaveBeenCalledWith(`formData${expected}`);
  });

  test("generate Form submission error on all mandatory fields", () => {
    const handleSubmit = jest.fn();
    render(<ContactForm />);
    const name = screen.getByPlaceholderText("Name");
    const email = screen.getByPlaceholderText("Email Address");
    const company = screen.getByPlaceholderText("Company Name");
    const title = screen.getByPlaceholderText("Title");
    const message = screen.getByPlaceholderText("Message");

    const submitButton = screen.getByText("submit");

    fireEvent.blur(name);
    fireEvent.blur(email);
    fireEvent.blur(company);
    fireEvent.blur(title);
    fireEvent.blur(message);

    fireEvent.click(submitButton);

    const generalError = screen.getAllByText(/This field is required/);
    expect(handleSubmit).not.toHaveBeenCalled();
    expect(generalError.length).toEqual(5);
  });

  test("generate Form submission error on invalid email", () => {
    render(<ContactForm />);
    const email = screen.getByPlaceholderText("Email Address");

    fireEvent.change(email, { target: { value: "test" } });
    fireEvent.blur(email);

    const emailError = screen.getByText(/Invalid email address/);
    expect(emailError).toBeInTheDocument();
  });

  test("generate Form submission error on invalid message length", () => {
    render(<ContactForm />);
    const message = screen.getByPlaceholderText("Message");

    fireEvent.change(message, { target: { value: "test message" } });
    fireEvent.blur(message);

    const messageError = screen.getByText(
      /Message must be at least 25 characters/
    );
    expect(messageError).toBeInTheDocument();
  });
});
