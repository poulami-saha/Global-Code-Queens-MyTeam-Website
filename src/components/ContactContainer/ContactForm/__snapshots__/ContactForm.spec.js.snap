// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormComponent renders ContactForm correctly and matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="flexColumnStart"
  >
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Name"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Email Address"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Company Name"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Title"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <textarea
        class=" false textarea"
        placeholder="Message"
      />
    </div>
    <button
      class="button disabled buttonSecondary"
      disabled=""
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`FormComponent submits Contact Form with Data 1`] = `
<DocumentFragment>
  <form
    class="flexColumnStart"
  >
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Name"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Email Address"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Company Name"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Title"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <textarea
        class=" false textarea"
        placeholder="Message"
      />
    </div>
    <button
      class="button disabled buttonSecondary"
      disabled=""
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`renders ContactForm correctly and matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="flexColumnStart"
  >
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Name"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Email Address"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Company Name"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Title"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <textarea
        class=" false textarea"
        placeholder="Message"
      />
    </div>
    <button
      class="button disabled buttonSecondary"
      disabled=""
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`submits Contact Form with Data 1`] = `
<DocumentFragment>
  <form
    class="flexColumnStart"
  >
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Name"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Email Address"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Company Name"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <input
        class=" false input"
        placeholder="Title"
        type="text"
        value=""
      />
    </div>
    <div
      class="formElement"
    >
      <textarea
        class=" false textarea"
        placeholder="Message"
      />
    </div>
    <button
      class="button disabled buttonSecondary"
      disabled=""
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;
